go/vt/vtqueryserver/endtoend_test.go

```diff
@@ -0,0 +1,314 @@
+							{
+								Name:           "0",
+								DbNameOverride: "vttest",
+							},
+						},
+					},
+				},
+			},
+			OnlyMySQL: true,
+		}
+		if err := cfg.InitSchemas("vttest", testSchema, nil); err != nil {
+			fmt.Fprintf(os.Stderr, "InitSchemas failed: %v\n", err)
+			return 1
+		}
+		defer os.RemoveAll(cfg.SchemaDir)
+		cluster := vttest.LocalCluster{
+			Config: cfg,
+		}
+		if err := cluster.Setup(); err != nil {
+			fmt.Fprintf(os.Stderr, "could not launch mysql: %v\n", err)
+			return 1
+		}
+		defer cluster.TearDown()
+
+		mysqlConnParams = cluster.MySQLConnParams()
+
+		proxySock := path.Join(cluster.Env.Directory(), "mysqlproxy.sock")
+
+		proxyConnParams.UnixSocket = proxySock
+		proxyConnParams.Uname = "proxy"
+		proxyConnParams.Pass = "letmein"
+
+		*mysqlServerSocketPath = proxyConnParams.UnixSocket
+		*mysqlAuthServerImpl = "none"
+
+		dbcfgs := dbconfigs.DBConfigs{
+			App: mysqlConnParams,
+		}
+
+		var err error
+		queryServer, err = initProxy(&dbcfgs)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "could not start proxy: %v\n", err)
+			return 1
+		}
+		defer queryServer.StopService()
+
+		initMySQLProtocol()
+		defer shutdownMySQLProtocol()
+
+		return m.Run()
+	}()
+	os.Exit(exitCode)
+}
+
+var testSchema = `
+create table test(id int, val varchar(256), primary key(id));
+create table valtest(intval int default 0, floatval float default null, charval varchar(256) default null, binval varbinary(256) default null, primary key(intval));
+`
+
+func testFetch(t *testing.T, conn *mysql.Conn, sql string, expectedRows int) *sqltypes.Result {
+	t.Helper()
+
+	result, err := conn.ExecuteFetch(sql, 1000, false)
+	if err != nil {
+		t.Errorf("error: %v", err)
+	}
+
+	if len(result.Rows) != expectedRows {
+		t.Errorf("expected %d rows but got %d", expectedRows, len(result.Rows))
+	}
+
+	return result
+}
+
+func testDML(t *testing.T, conn *mysql.Conn, sql string, expectedNumQueries int64, expectedRowsAffected uint64) {
+	t.Helper()
+
+	numQueries := tabletenv.MySQLStats.Count()
+	result, err := conn.ExecuteFetch(sql, 1000, false)
+	if err != nil {
+		t.Errorf("error: %v", err)
+	}
+	numQueries = tabletenv.MySQLStats.Count() - numQueries
+
+	if numQueries != expectedNumQueries {
+		t.Errorf("expected %d mysql queries but got %d", expectedNumQueries, numQueries)
+	}
+
+	if result.RowsAffected != expectedRowsAffected {
+		t.Errorf("expected %d rows affected but got %d", expectedRowsAffected, result.RowsAffected)
+	}
+}
+
+func TestQueries(t *testing.T) {
+	ctx := context.Background()
+	conn, err := mysql.Connect(ctx, &proxyConnParams)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Try a simple query case.
+	testFetch(t, conn, "select * from test", 0)
+
+	// Try a simple error case.
+	_, err = conn.ExecuteFetch("select * from aa", 1000, true)
+	if err == nil || !strings.Contains(err.Error(), "table aa not found in schema") {
+		t.Fatalf("expected error but got: %v", err)
+	}
+}
+
+func TestAutocommitDMLs(t *testing.T) {
```

Comment:
I couldn't find a test for autocommit=0 -> autocommit=1 transition (where it needs to commit the current transaction). Ref: https://github.com/youtube/vitess/blob/master/go/vt/vtgate/executor_test.go#L348